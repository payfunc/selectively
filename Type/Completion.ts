import { Cursor } from "./Cursor"

export interface Completion {
	value: string
	cursor?: Cursor | number
}
export namespace Completion {
	export function prepend(prefix: string, completion: Completion, suffix?: string): Completion
	export function prepend(prefix: string, completions: Completion[], suffix?: string): Completion[]
	export function prepend(
		prefix: string,
		completion: Completion | Completion[],
		suffix?: string
	): Completion | Completion[] {
		return Array.isArray(completion)
			? completion.map(c => prepend(prefix, c, suffix))
			: {
					value: prefix + completion.value + (suffix ?? ""),
					cursor:
						typeof completion?.cursor == "number"
							? completion?.cursor + prefix.length
							: typeof completion?.cursor == "object"
							? { start: completion.cursor.start + prefix.length, end: completion.cursor.end + prefix.length }
							: undefined,
			  }
	}
}
