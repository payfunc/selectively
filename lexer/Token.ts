import { Error } from "@cogneco/mend"

export interface Token {
	value: string
	region?: Error.Region
}

export namespace Token {
	export function isSymbol(token: Token) {
		return ["!", "(", ")", "[", "]", "|", "*", ":", ".", "<=", ">=", "<", ">"].some(s => s == token.value)
	}
}
