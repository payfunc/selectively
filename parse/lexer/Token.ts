import { Error } from "@cogneco/mend"

export interface Token {
	value: string | number
	region: Error.Region
}
