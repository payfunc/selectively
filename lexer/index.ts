import { Error, IO, Utilities } from "@cogneco/mend"
import { Source } from "./Source"
import { Token } from "./Token"

function tokenize(reader: IO.Reader | string, errorHandler?: Error.Handler): Utilities.Enumerator<Token> {
	const source = new Source(reader, errorHandler)
	return new Utilities.Enumerator<Token>(() => {
		let result: Token | undefined
		let peekSymbolLength: number | false = 1
		if (!source.isEmpty) {
			while (source.peekIsWhitespace() && !source.peekIsSymbol())
				source.read()
			if ((peekSymbolLength = source.peekIsSymbol()))
				result = { value: source.read(peekSymbolLength) || "", region: source.mark() }
			else {
				let value = ""
				while (!source.isEmpty && !source.peekIsWhitespace() && !source.peekIsSymbol())
					value += source.read()
				if (value)
					result = { value, region: source.mark() }
			}
		}
		return result
	})
}

export { Source, Token, tokenize }
