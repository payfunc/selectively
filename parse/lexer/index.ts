import { Error, IO, Utilities } from "@cogneco/mend"
import { Source } from "./Source"
import { Token } from "./Token"

function tokenize(reader: IO.Reader | string, errorHandler?: Error.Handler): Utilities.Enumerator<Token> {
	const source = new Source(reader, errorHandler)
	return new Utilities.Enumerator<Token>(() => {
		let result: Token | undefined
		if (!source.isEmpty) {
			while (source.peekIsWhitespace())
				source.read()
			// if (source.peekIsDoubleSymbol())
			// 	result = { value: source.read(2) || "", region: source.mark() }
			if (source.peekIsSymbol())
				result = { value: source.read() || "", region: source.mark() }
			else {
				let value = ""
				while (!source.isEmpty && !source.peekIsWhitespace() && !source.peekIsSymbol())
					value += source.read()
				if (value)
					result = { value, region: source.mark() }
			}
		}
		return result
	})
}

export { Source, Token, tokenize }
