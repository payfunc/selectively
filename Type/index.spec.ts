import * as selectively from "../index"
import { Token } from "../lexer"
import { tokenize } from "../lexer"
import { Completion } from "./Completion"

function t(data: string): Token[] {
	return tokenize(data)
		.map(t => ({ value: t.value }))
		.toArray()
}
const testing = {
	merchant: {
		name: "Example AB",
		descriptor: "Example Company",
		country: "SE",
		currency: "SEK",
		captured: 1000,
		fees: 1000,
		refundable: 1000,
		settled: 1000,
		scheme: ["visa", "mastercard"],
	},
	authorization: {
		amount: 10,
		currency: "SEK",
		card: {
			iin: "411111",
			last4: "1111",
			scheme: "visa",
			csc: "present",
			type: "debit",
		},
		capture: "auto",
		descriptor: "Example AB",
		recurring: "initial",
		verification: "rejected",
	},
}
const output = selectively.Type.convert(testing)
describe("selectively.Type", () => {
	it("Tokenizer", () => {
		expect(t("")).toEqual([])
	})
	it("complete", () => {
		expect(Completion.stringify(output.complete(t("")))).toEqual(["merchant|", "authorization|"])
		expect(Completion.stringify(output.complete(t("merc")))).toEqual(["merchant|"])
		expect(Completion.stringify(output.complete(t("merchant")))).toEqual(["merchant.|", "merchant:|"])
		expect(Completion.stringify(output.complete(t("merchant.")))).toEqual([
			"merchant.name|",
			"merchant.descriptor|",
			"merchant.country|",
			"merchant.currency|",
			"merchant.captured|",
			"merchant.fees|",
			"merchant.refundable|",
			"merchant.settled|",
			"merchant.scheme|",
		])
		expect(Completion.stringify(output.complete(t("merchant.name")))).toEqual([
			"merchant.name:|",
			"merchant.name>|",
			"merchant.name>=|",
			"merchant.name<|",
			"merchant.name<=|",
		])
		expect(Completion.stringify(output.complete(t("merchant.name:")))).toEqual([
			"merchant.name:*|",
			"merchant.name:*|*",
			"merchant.name:/|/",
			"merchant.name:!|",
			"merchant.name:|*",
		])
	})

	it("right hand side test with string on left hand", () => {
		expect(Completion.stringify(output.complete(t("merchant.name>")))).toEqual([
			"merchant.name>merchant.|",
			"merchant.name>authorization.|",
			"merchant.name>=|",
		])
		expect(Completion.stringify(output.complete(t("merchant.name>authorization.")))).toEqual([
			"merchant.name>authorization.currency|",
			"merchant.name>authorization.card.|",
			"merchant.name>authorization.capture|",
			"merchant.name>authorization.descriptor|",
			"merchant.name>authorization.recurring|",
			"merchant.name>authorization.verification|",
		])
		expect(Completion.stringify(output.complete(t("merchant.name>authorization.car")))).toEqual([
			"merchant.name>authorization.card.|",
		])
		expect(Completion.stringify(output.complete(t("merchant.name>authorization.card.")))).toEqual([
			"merchant.name>authorization.card.iin|",
			"merchant.name>authorization.card.last4|",
			"merchant.name>authorization.card.scheme|",
			"merchant.name>authorization.card.csc|",
			"merchant.name>authorization.card.type|",
		])
	})
	it("right hand side test with number on left hand", () => {
		expect(Completion.stringify(output.complete(t("merchant.captured")))).toEqual([
			"merchant.captured>|",
			"merchant.captured>=|",
			"merchant.captured<|",
			"merchant.captured<=|",
			"merchant.captured:|",
		])
		expect(Completion.stringify(output.complete(t("merchant.captured>")))).toEqual([
			"merchant.captured>merchant.|",
			"merchant.captured>authorization.|",
			"merchant.captured>=|",
		])
		expect(Completion.stringify(output.complete(t("merchant.captured>auth")))).toEqual([
			"merchant.captured>authorization.|",
		])
		expect(Completion.stringify(output.complete(t("merchant.captured>authorization.")))).toEqual([
			"merchant.captured>authorization.amount|",
		])
		expect(Completion.stringify(output.complete(t("merchant.captured>authorization.amount")))).toEqual([
			"merchant.captured>authorization.amount + |",
			"merchant.captured>authorization.amount - |",
			"merchant.captured>authorization.amount * |",
		])
		expect(Completion.stringify(output.complete(t("merchant.captured>authorization.amount + ")))).toEqual([
			"merchant.captured>authorization.amount + merchant.|",
			"merchant.captured>authorization.amount + authorization.|",
		])
	})
})
