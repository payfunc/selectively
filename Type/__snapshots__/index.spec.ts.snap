// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectively.Type Array 1`] = `
Array [
  Object {
    "cursor": 7,
    "value": ".every()",
  },
]
`;

exports[`selectively.Type Array 2`] = `
Array [
  Object {
    "cursor": 7,
    "value": ".every()",
  },
]
`;

exports[`selectively.Type Boolean 1`] = `
Boolean {
  "class": "boolean",
}
`;

exports[`selectively.Type Boolean 2`] = `
Array [
  Object {
    "cursor": 5,
    "value": ".true",
  },
]
`;

exports[`selectively.Type Boolean 3`] = `
Array [
  Object {
    "cursor": 5,
    "value": ".true",
  },
  Object {
    "cursor": 6,
    "value": ".false",
  },
]
`;

exports[`selectively.Type Object 1`] = `
TObject {
  "class": "object",
  "completions": Array [
    Object {
      "value": "name",
    },
  ],
  "properties": Object {
    "name": String {
      "class": "string",
      "completion": Object {
        "value": "",
      },
      "value": "",
    },
  },
}
`;

exports[`selectively.Type Tokenizer 1`] = `
Array [
  Object {
    "value": ".",
  },
  Object {
    "value": "object",
  },
  Object {
    "value": ".",
  },
  Object {
    "value": "has",
  },
  Object {
    "value": "(",
  },
  Object {
    "value": ")",
  },
]
`;

exports[`selectively.Type Tokenizer 2`] = `
Array [
  Object {
    "value": ".",
  },
  Object {
    "value": "*",
  },
  Object {
    "value": "*",
  },
]
`;

exports[`selectively.Type complete 1`] = `
Array [
  Object {
    "cursor": 3,
    "value": ".id",
  },
  Object {
    "cursor": 7,
    "value": ".status",
  },
  Object {
    "cursor": 5,
    "value": ".has()",
  },
]
`;

exports[`selectively.Type complete 2`] = `
Array [
  Object {
    "cursor": 7,
    "value": ".has(id)",
  },
]
`;

exports[`selectively.Type complete 3`] = `
Array [
  Object {
    "cursor": 7,
    "value": ".has(id)",
  },
  Object {
    "cursor": 11,
    "value": ".has(status)",
  },
]
`;

exports[`selectively.Type complete 4`] = `
Array [
  Object {
    "cursor": 5,
    "value": ".has()",
  },
]
`;

exports[`selectively.Type complete 5`] = `
Array [
  Object {
    "cursor": 5,
    "value": ".has()",
  },
]
`;

exports[`selectively.Type complete 6`] = `
Array [
  Object {
    "cursor": 4,
    "value": ".id.",
  },
]
`;

exports[`selectively.Type complete 7`] = `
Array [
  Object {
    "cursor": 9,
    "value": ".id.first",
  },
  Object {
    "cursor": 10,
    "value": ".id.second",
  },
  Object {
    "cursor": 8,
    "value": ".id.has()",
  },
]
`;

exports[`selectively.Type complete 8`] = `
Array [
  Object {
    "cursor": 13,
    "value": ".id.has(first)",
  },
  Object {
    "cursor": 14,
    "value": ".id.has(second)",
  },
]
`;

exports[`selectively.Type complete 9`] = `
Array [
  Object {
    "cursor": 9,
    "value": ".id.first",
  },
]
`;

exports[`selectively.Type complete 10`] = `
Array [
  Object {
    "cursor": 10,
    "value": ".id.first:",
  },
  Object {
    "cursor": 10,
    "value": ".id.first>",
  },
  Object {
    "cursor": 11,
    "value": ".id.first>=",
  },
  Object {
    "cursor": 10,
    "value": ".id.first<",
  },
  Object {
    "cursor": 11,
    "value": ".id.first<=",
  },
]
`;

exports[`selectively.Type complete 11`] = `
Array [
  Object {
    "cursor": 14,
    "value": ".id.first<TODO",
  },
  Object {
    "cursor": 11,
    "value": ".id.first<=",
  },
]
`;

exports[`selectively.Type complete 12`] = `
Array [
  Object {
    "cursor": 11,
    "value": ".id.first:*",
  },
  Object {
    "cursor": 11,
    "value": ".id.first:**",
  },
]
`;

exports[`selectively.Type complete 13`] = `
Array [
  Object {
    "cursor": 15,
    "value": ".id.first:*test",
  },
  Object {
    "cursor": 12,
    "value": ".id.first:**",
  },
]
`;

exports[`selectively.Type complete 14`] = `
Array [
  Object {
    "cursor": 15,
    "value": ".id.first:*test*",
  },
]
`;

exports[`selectively.Type complete 15`] = `
Array [
  Object {
    "cursor": 14,
    "value": ".id.first:test*",
  },
]
`;

exports[`selectively.Type complete 16`] = `Array []`;

exports[`selectively.Type complete 17`] = `
Array [
  Object {
    "cursor": 15,
    "value": ".id.first:*test*",
  },
]
`;

exports[`selectively.Type complete 18`] = `Array []`;

exports[`selectively.Type complete 19`] = `
Array [
  Object {
    "cursor": 15,
    "value": ".id.first:*test",
  },
]
`;

exports[`selectively.Type complete 20`] = `Array []`;

exports[`selectively.Type complete 21`] = `
Array [
  Object {
    "cursor": 20,
    "value": ".status.statusArray.",
  },
]
`;

exports[`selectively.Type complete 22`] = `
Array [
  Object {
    "cursor": 26,
    "value": ".status.statusArray.every()",
  },
]
`;

exports[`selectively.Type complete 23`] = `
Array [
  Object {
    "cursor": 26,
    "value": ".status.statusArray.every()",
  },
]
`;

exports[`selectively.Type complete 24`] = `
Array [
  Object {
    "cursor": 26,
    "value": ".status.statusArray.every()",
  },
]
`;

exports[`selectively.Type complete 25`] = `
Array [
  Object {
    "cursor": 30,
    "value": ".status.statusArray.every(TODO)",
  },
]
`;

exports[`selectively.Type complete 26`] = `31`;
